cpmpy/solvers/z3.py

```diff
@@ -299,6 +299,11 @@ def _z3_expr(self, cpm_con, reify=False):
                 return z3.Or(self._z3_expr(cpm_con.args))
             elif cpm_con.name == '->':
                 return z3.Implies(*self._z3_expr(cpm_con.args, reify=True))
+            elif cpm_con.name == 'not':
+                if is_boolexpr(cpm_con.args[0]):
+                    return z3.Not(self._z3_expr(cpm_con.args[0]))
+                else:
+                    return self._z3_expr(cpm_con.args[0] == 0)
```

Comment:
we, cpmpy, should never allow a not expression around a non-boolexpr, so that this case can never trigger